suitcase/tiff/__init__.py

```diff
@@ -1,25 +1,160 @@
 # NotImplementError, so that a client importing this library can immediately
 # know which portions of the suitcase API it supports without calling any
 # functions.
 #
-# def export(...)
-#     ...
-#
-#
-# def ingest(...):
-#     ...
-#
-#
-# def reflect(...):
-#     ...
-#
-#
-# handlers = []
+from collections import defaultdict
+from pathlib import Path
+import json
+import tifffile
+import event_model
+import numpy
+import suitcase
+from ._version import get_versions
+
+__version__ = get_versions()['version']
+del get_versions
+
+
+class SuitcaseTiffError(Exception):
+    ...
+
+
+class NonSupportedDataShape(SuitcaseTiffError):
+    '''used to indicate that non-supported data type is being saved
+    '''
+    ...
+
+
+def export(gen, directory, file_prefix='', **kwargs):
+    """
+    Export a stream of documents to tiff file(s) and one JSON file of metadata.
+
+    Creates {filepath}_meta.json and then {filepath}_{stream_name}.tiff
+    for every Event stream.
+
+    The structure of the json is::
+
+            {'metadata': {'start': start_doc, 'stop': stop_doc,
+                          'descriptors': {stream_name1: 'descriptor',
+                                          stream_name2: ...}},
+             stream_name1: {'seq_num': [], 'uid': [], 'time': [],
+                         'timestamps': {det_name1:[], det_name2:[],...},
+             stream_name2: ...}}
+
+            .. note::
+
+                This schema was chosen as the API is very similar to the
+                intake-databroker API. The same schema is used for all json
+                files created with our base suitcase export functions.
+
+    Parameters
+    ----------
+    gen : generator
+        expected to yield (name, document) pairs
+
+    directory : string, Path or Wrapper
+        The filepath and filename suffix to use in the output files or a file
+        handle factory wrapper(see ADD LINK HERE). An empty string will place
+        the file in the current directory.
+
+    file_prefix : str
+        An optional prefix for the file names that will be created, default is
+        an empty string.
+
+    **kwargs : kwargs
+        kwargs to be passed to tifffile.TiffWriter.save.
+
+    Returns
+    -------
+    dest : tuple
+        filepaths of generated files
+    """
+    meta = defaultdict(dict)  # to be exported as JSON at the end
+    meta['metadata']['descriptors'] = defaultdict(dict)
+    stream_names = {}  # dict to capture stream_names for each descriptor uid
+    files = {}  # map descriptor uid to file handle of tiff file
+    filenames = {}  # map descriptor uid to file names of tiff files
+
+    # Load up the correct wrapper.
+    if isinstance(directory, (str, Path)):
+        wrapper = suitcase.utils.MultiFileWrapper(directory)
+        directory = Path
+    else:
+        wrapper = directory
+
+    # set file_joiner to `_` if file_prefix is not empty.
+    if file_prefix:
```

Comment:
I do not think there should be a joiner,  as currently coded there is no way to ask for a filename that looks like `prefix-postfix.tif` which seems like a reasonable thing to ask for.